Add documentation for how OpenCL runtime manages buffers (#81)
intel/fpga-runtime-for-opencl: docs/buffers.md

Some other note about memory operation during enqueue:
1. Device local pointer size is 4, device global pointer size is always the device's address bit integer division 8

Before [submitting kernel](https://github.com/intel/fpga-runtime-for-opencl/blob/1264543c0361530f5883e35dc0c9d48ac0fd3653/src/acl_kernel.cpp#L2982-L3093) to device queue. It first checks if the device is programmed, if not, it will [queue reprogram device operation](https://github.com/intel/fpga-runtime-for-opencl/blob/1264543c0361530f5883e35dc0c9d48ac0fd3653/src/acl_kernel.cpp#L3034) to do so. Then it will [arrange memory migration](https://github.com/intel/fpga-runtime-for-opencl/blob/1264543c0361530f5883e35dc0c9d48ac0fd3653/src/acl_kernel.cpp#L3043) for each kernel memory argument.

Different from the device operation resulting from enqueue read/write, the memory migration calls on [`acl_mem_migrate_buffer`](https://github.com/intel/fpga-runtime-for-opencl/blob/b08e0af97351718ce0368a9ee507242b35f4929e/src/acl_mem.cpp#L5412-L5665) (i.e memory transfer and memory migration behave differently). 

> In memory migration:

Reviewer: shouldn't this also be a section ?

Author: Thanks for the approval! I have wrapped memory migration into a subsection.

A synopsis of buffer use is as follows:

```cpp
#define GMEM_DATA_VEC 8

cl_platform_id platform;
cl_int err;
// General setups
err = clGetPlatformIDs(1, &platform, NULL);
err = clGetDeviceIDs(platform, CL_DEVICE_TYPE_ALL, 1, &device, NULL);
cl_context context = clCreateContext(NULL, 1, &device, NULL, NULL, &err);
cl_command_queue queue = clCreateCommandQueue(context, device, CL_QUEUE_PROFILING_ENABLE, &err);
// aocx_size is the size of aocx file
// aocx_contents is the content of aocx file
cl_program program = clCreateProgramWithBinary(context, 1, &device, &aocx_size, (const unsigned char **)&aocx_contents, &binary_status, &err);
cl_kernel kernel = clCreateKernel(program, "kernel_name", &err);

// Host side data that contains the information needed for kernel execution
unsigned int *host_data = (unsigned int *)acl_aligned_malloc(global_work_size*sizeof(unsigned int)*GMEM_DATA_VEC);

// Set up the expectation for a buffer, but not actually allocating memory yet
cl_mem mem = clCreateBuffer(context, CL_MEM_READ_WRITE, global_work_size*sizeof(unsigned int)*GMEM_DATA_VEC, NULL, &err);

// Enqueue a memory transfer operation

> cl_event event = {};

Author: Thank you for all the suggestions!! I have made the changes accordingly to other places as well.

err = clEnqueueWriteBuffer(queue, mem, CL_FALSE, 0, global_work_size*sizeof(unsigned int)*GMEM_DATA_VEC, host_data, 0, NULL, &event);

// Informs the kernel that one of its arguments is the buffer
err = clSetKernelArg(kernel, 0, sizeof(cl_mem), &mem);

// Submit kernel execution to queue
err = clEnqueueNDRangeKernel(queue, kernel, 2, NULL, &global_work_size_2d[0], &local_work_size_2d[0], 0, NULL, &event);

... // Block on kernel finish and clean up resources
```

## Summary

> * Buffers are not allocated on the device until the kernel executes, the `clEnqueueWriteBuffer` function is called, or `CL_MEM_COPY_HOST_PTR` flag is defined when calling the function `clCreateBuffer`.

Reviewer 2: Swapping the order preserves the flow:
```suggestion
* Buffers are not allocated on the device until the kernel executes, the `clEnqueueWriteBuffer` function is called, or the `clCreateBuffer` function is called with the `CL_MEM_COPY_HOST_PTR` flag.
```

* Buffer handles memory migration between the host and device automatically.

> * SYCL* runtime typically calls OpenCL functions in a different order in comparison to OpenCL's preferred order. Specifically, SYCL* runtime calls functions in the order of `clCreateBuffer` > `clEnqueueWriteBuffer` > `clSetKernelArg` > `clEnqueueNDRangeKernel`. In OpenCL, the optimal calls are `clCreateBuffer` > `clSetKernelArg` > `clEnqueueWriteBuffer` >  `clEnqueueNDRangeKernel`. The difference in order of calls sometimes leads to an unexpected result (eg. extra host to device memory copy), but you can avoid it if the `clSetKernelArg` function does not modify `cl_mem` attributes.

Reviewer 2: Does the latter preference actually apply to OpenCL and (most of) its implementations, or only to the FPGA Runtime for OpenCL?

Reviewer 2: This item is quite long and the order of calls cumbersome to read. How about removing the entire sentence "Specifically …" in favour of a link to the section below? Then expand the section below to include a second list of calls in the order preferred by the FPGA Runtime for OpenCL.

* The runtime keeps track of which memory addresses are occupied and decides where the next allocation should be. Specifically, it uses a first-fit allocation algorithm.
* The actual memory allocation is made with MMD calls by passing the device address and size.
* Device address has a special encoding, which is the same between buffer and USM pointers.
* Even if you do not explicitly transfer the memory before launching the kernel, the memory is still migrated right before the kernel executes (i.e., `clEnqueueWriteBuffer` is not necessary).

## SYCL Runtime Flow

SYCL* runtime calls the memory operation in the following order, which are explained in the subsections below:
1. `clCreateBuffer`
2. `clEnqueueWriteBuffer`
3. `clSetKernelArg`

> 4. `clEnqueueKernel` / Task

Reviewer 2: ```suggestion
4. `clEnqueueKernel` / `clEnqueueTask`
```

> **NOTE:** SYCL* runtime calls are not the preferred order from the OpenCL runtime perspective. However, SYCL runtime* cannot change the order of API invocations as it needs to handle other vendors' requirements. Not all vendors can get away with setting kernel argument first before enqueuing write buffer (As discussed in one of the [issues](https://github.com/intel/llvm/discussions/4627)).

Reviewer 2: OpenCL runtime → FPGA Runtime for OpenCL

Reviewer 2: (As discussed in one of the [issues](https://github.com/intel/llvm/discussions/4627)) → (see [discussion](https://github.com/intel/llvm/discussions/4627))

   * [`copy_globalmem_to_globalmem`](https://github.com/intel/fpga-runtime-for-opencl/blob/fc99b92704a466f7dc4d84bd45d465d64d03dbb0/src/acl_hal_mmd.cpp#L1763-L1873) - If the source and destination are on the same device, then runtime directly calls the MMD function [`aocl_mmd_copy`](https://gitlab.devtools.intel.com/OPAE/opencl-bsp/-/blob/master/agilex_f_dk/source/host/ccip_mmd.cpp#L881-891). Otherwise, it uses both [`aocl_mmd_read`](https://gitlab.devtools.intel.com/OPAE/opencl-bsp/-/blob/master/agilex_f_dk/source/host/ccip_mmd.cpp#L870-879) and [`aocl_mmd_write`](https://gitlab.devtools.intel.com/OPAE/opencl-bsp/-/blob/master/agilex_f_dk/source/host/ccip_mmd.cpp#L870-879) functions to copy from the source device to host and then from the host to the destination device. All operations are blocking. The runtime keeps calling MMD's `yield` (sleep function) function until read and write operations are completed.

### `clSetKernelArg`: What if the `clEnqueueWriteBuffer` function is not called?

> When the `clEnqueueWriteBuffer` function is not called, memory transfer automatically happens before launching the kernel that uses it. There is an enqueued memory transfer device operation before every kernel launch device operation. The only difference between calling and not calling the `enqueueWriteBuffer` function is whether the enqueued memory transfer actually copies the memory. 

Reviewer 2: that uses it → that uses the buffer
